### PR TITLE
RUBY-1222 Use IP address for creating sockets and stop caching hostna…

### PR DESCRIPTION
**Pull request by saghm**
…me resolution

This pull request fixes two bugs in the hostname resolution/socket creation functionality of the driver. The first is that sockets are incorrectly being created with the hostname rather than the IP address that was obtained with `getaddrinfo`. This breaks the ability to fallback to IPv4 in the case that connecting to the IPv6 address fails. The second bug is the caching of the resolver in the address; doing a DNS lookup each time a socket is created ensures that the driver will detect any changes to where the hostname resolves to without requiring the Client to be recreated.

### CODE REVIEW

**Comment by jarthod** [comment]
Looks good to me :+1: wouldn't this deserve some tests to avoid future another regression?

File: lib/mongo/address.rb [hunk 1]
```diff
@@ -146,7 +152,7 @@ def inspect
     #
     # @since 2.0.0
     def socket(socket_timeout, ssl_options = {})
-      @resolver ||= initialize_resolver!(ssl_options)
+      @resolver = initialize_resolver!(ssl_options)
       @resolver.socket(socket_timeout, ssl_options)
```
Comment (p-mongo) [inline lib/mongo/address.rb]: Since the resolver is now local, it should no longer be an instance variable (remove `@`).

Comment (saghm) [reply lib/mongo/address.rb]: Hmm, looks like some tests fail when I do that. I'll look into what it takes to fix them.

**Comment by saghm** [comment]
Hi @jarthod! Sorry for the delay in the response about testing. It looks like the mock test that existed to test this fallback only tested that the socket was able to make a connection in the case that both IPv6 and IPv4 were returned by `getaddrinfo`; I've changed this to assert that the socket actually resolves to an IPv4 address (see [this commit](https://github.com/mongodb/mongo-ruby-driver/pull/1058/commits/c951d40be49a679ade3246c6526e9f548a50d80b)). The test fails on master but passes on the branch for this PR, so I think this should be sufficient to detect any failures in the future.

**Comment by saghm** [comment]
@p-mongo Sorry for all the work on this after the initial LGTM! I think this has everything it should need now, and it's passing with the modified tests, so only one more look-over needed

**Pull request merged**